backend/layers/processing/process_validate_atac.py

```diff
@@ -0,0 +1,224 @@
+        view = memoryview(buffer)
+        with open(file_name, "rb") as f:
+            while chunk := f.readinto(buffer):
+                hashobj.update(view[:chunk])
+        return hashobj.hexdigest()
+
+    def process(
+        self,
+        collection_version_id: CollectionVersionId,
+        dataset_version_id: DatasetVersionId,
+        manifest: IngestionManifest,
+        datasets_bucket: str,
+    ):
+        """
+
+        :param collection_version_id:
+        :param dataset_version_id:
+        :param manifest:
+        :param datasets_bucket:
+        :return:
+        """
+        # Download the original dataset files from URI
+        local_anndata_filename = self.download_from_source_uri(
+            source_uri=str(manifest.anndata),
+            local_path=CorporaConstants.ORIGINAL_H5AD_ARTIFACT_FILENAME,
+        )
+
+        if self.skip_atac_validation(local_anndata_filename, manifest, dataset_version_id):
+            return
+
+        # Download the original fragment file from URI
+        local_fragment_filename = self.download_from_source_uri(
+            source_uri=str(manifest.atac_fragment), local_path=CorporaConstants.ORIGINAL_ATAC_FRAGMENT_FILENAME
+        )
+
+        # Hash the original fragment file. This is used to check if the new fragment file is the same as the old
+        # fragment file to avoid uploading the same file multiple times
+        if self.business_logic.is_already_ingested(manifest.atac_fragment):
+            original_fragment_hash = self.hash_file(local_fragment_filename)
+        else:
+            # new fragment file is from a private source, so we must upload it
+            original_fragment_hash = None
+
+        # Validate the fragment with anndata file
+        try:
+            errors = self.schema_validator.validate_atac(local_fragment_filename, local_anndata_filename)
+        except Exception as e:
+            # for unexpected errors, log the exception and raise a ValidationAtacFailed exception
+            self.logger.exception("validation failed")
+            self.update_processing_status(
+                dataset_version_id, DatasetStatusKey.ATAC_FRAGMENT, DatasetConversionStatus.FAILED
+            )
+            raise ValidationAtacFailed(errors=[str(e)]) from None
+
+        if errors:
+            # if the validation fails, update the processing status and raise a ValidationAtacFailed exception
+            self.update_processing_status(
+                dataset_version_id, DatasetStatusKey.ATAC_FRAGMENT, DatasetConversionStatus.FAILED
+            )
+            raise ValidationAtacFailed(errors=errors)
+
+        # check to see if the new fragments is the same as the old fragment
+        # if it is the same, skip the upload and use link the old fragment to the new dataset
+        # This case is for when the dataset is reprocessed, or updated
+        if original_fragment_hash is not None and original_fragment_hash == self.hash_file(local_fragment_filename):
```

Comment:
Why do we need to take the hashes and compare them at all if original_fragment_hash is only populated if we confirm that the given uri is already ingested?

Reply:
because the CLI might change which may process the data differently.

Reply:
I see. In that case, shouldn't we be comparing the TBI hashes rather than the TSV? since the TSV is pointing to the same file, its the TBI that could be generated differently.

Reply:
could be misunderstanding something here

Reply:
also, to consider--for very large files, reading from disk again would make this a pretty slow operation and cost us in AWS batch

Reply:
No the files will only be hashed during a migration.